### PR TITLE
Change maxPriorityFeePerGas, to fix issue with Besu

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

File: src/common/execution.py [hunk 1]
```diff
@@ -104,8 +105,29 @@ async def get_last_rewards_update() -> RewardVoteInfo | None:
...
+async def calculate_median_priority_fee(block_id='latest') -> Wei:
+    try:
+        latest = await execution_client.eth.get_block(block_id)
+
+        # collect maxPriorityFeePerGas for all transactions in the block
+        priority_fees = []
+        for tx_hash in latest.transactions:
+            tx = await execution_client.eth.get_transaction(tx_hash)
+            if 'maxPriorityFeePerGas' in tx:
+                print(tx.maxPriorityFeePerGas)
```
Comment (tsudmi) [inline src/common/execution.py:124]: redundant

File: src/common/execution.py [hunk 2]
```diff
@@ -104,8 +105,29 @@ async def get_last_rewards_update() -> RewardVoteInfo | None:
...
+    priority_fee = await calculate_median_priority_fee()  # type: ignore
     latest_block = await execution_client.eth.get_block('latest')  # type: ignore
     base_fee = latest_block['baseFeePerGas']
     max_fee_per_gas = priority_fee + 2 * base_fee
     return Wei(max_fee_per_gas)
+
+
+async def calculate_median_priority_fee(block_id='latest') -> Wei:
+    try:
+        latest = await execution_client.eth.get_block(block_id)
```
Comment (tsudmi) [inline src/common/execution.py:117]: rename to block

File: src/common/execution.py [hunk 3]
```diff
@@ -104,8 +105,28 @@ async def get_last_rewards_update() -> RewardVoteInfo | None:
...
 async def get_max_fee_per_gas() -> Wei:
-    priority_fee = await execution_client.eth.max_priority_fee  # type: ignore
+    priority_fee = await calculate_median_priority_fee()  # type: ignore
     latest_block = await execution_client.eth.get_block('latest')  # type: ignore
     base_fee = latest_block['baseFeePerGas']
     max_fee_per_gas = priority_fee + 2 * base_fee
     return Wei(max_fee_per_gas)
+
+
+async def calculate_median_priority_fee(block_id='latest') -> Wei:
```
Comment (tsudmi) [inline src/common/execution.py:115]: please add argument type

File: src/common/execution.py [hunk 4]
```diff
@@ -104,8 +105,28 @@ async def get_last_rewards_update() -> RewardVoteInfo | None:
...
+            print(tx.maxPriorityFeePerGas)
+            priority_fees.append(tx.maxPriorityFeePerGas)
+
+    if not priority_fees:
+        return await calculate_median_priority_fee(latest['number'] - 1)
+
+    # calculate the median priority fee
+    median_priority_fee = statistics.median(priority_fees)
+
+    return Wei(median_priority_fee)
```
Comment (tsudmi) [inline src/common/execution.py:132]:
```suggestion
    return Wei(statistics.median(priority_fees))
```

File: src/common/execution.py [hunk 5]
```diff
@@ -104,8 +105,24 @@ async def get_last_rewards_update() -> RewardVoteInfo | None:
 
 
 async def get_max_fee_per_gas() -> Wei:
-    priority_fee = await execution_client.eth.max_priority_fee  # type: ignore
+    priority_fee = await calculate_median_priority_fee()  # type: ignore
```
Comment (evgeny-stakewise) [inline src/common/execution.py:108]: I see a lot of queries to execution node inside  `calculate_median_priority_fee`. Maybe try calling `max_priority_fee` as it was before. Then in the case of failure try more complicated way.

File: src/common/execution.py [hunk 6]
```diff
@@ -104,8 +105,24 @@ async def get_last_rewards_update() -> RewardVoteInfo | None:
...
+
+    # collect maxPriorityFeePerGas for all transactions in the block
+    priority_fees = []
+    for tx_hash in block.transactions:
+        tx = await execution_client.eth.get_transaction(tx_hash)
+        if 'maxPriorityFeePerGas' in tx:
+            priority_fees.append(tx.maxPriorityFeePerGas)
+
+    if not priority_fees:
+        return await calculate_median_priority_fee(block['number'] - 1)
```
Comment (evgeny-stakewise) [inline src/common/execution.py:126]: (optional) recursion is not necessary here. The same thing can be done with a loop.

File: src/common/execution.py [hunk 7]
```diff
@@ -104,8 +106,27 @@ async def get_last_rewards_update() -> RewardVoteInfo | None:
...
+    except MethodUnavailable:
+        priority_fee = await calculate_median_priority_fee()  # type: ignore
     latest_block = await execution_client.eth.get_block('latest')  # type: ignore
     base_fee = latest_block['baseFeePerGas']
     max_fee_per_gas = priority_fee + 2 * base_fee
     return Wei(max_fee_per_gas)
+
+
+async def calculate_median_priority_fee(block_id: str = 'latest') -> Wei:
+    block = await execution_client.eth.get_block(block_id)
```
Comment (evgeny-stakewise) [inline src/common/execution.py:120]: retries?

File: src/common/execution.py [hunk 8]
```diff
@@ -104,8 +106,27 @@ async def get_last_rewards_update() -> RewardVoteInfo | None:
...
+
+    # collect maxPriorityFeePerGas for all transactions in the block
+    priority_fees = []
+    for tx_hash in block.transactions:
+        tx = await execution_client.eth.get_transaction(tx_hash)
```
Comment (evgeny-stakewise) [inline src/common/execution.py:125]: retries?

File: src/common/execution.py [hunk 9]
```diff
@@ -104,8 +107,32 @@ async def get_last_rewards_update() -> RewardVoteInfo | None:
...
     return Wei(max_fee_per_gas)
+
+
+async def calculate_median_priority_fee(block_id: str = 'latest') -> Wei:
+    block = await execution_client.eth.get_block(block_id)
+
+    # collect maxPriorityFeePerGas for all transactions in the block
+    priority_fees = []
+    for tx_hash in block.transactions:
+        tx = await execution_client.eth.get_transaction(HexStr(tx_hash))
```
Comment (evgeny-stakewise) [inline src/common/execution.py:126]:
```suggestion
        tx = await eth_get_transaction(HexStr(tx_hash))
```

File: src/common/execution.py (hunk 7 above)
Comment (evgeny-stakewise) [reply src/common/execution.py:121]: Still no retries here

**Review by evgeny-stakewise — APPROVED** [review]

**Pull request merged**